spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -1005,45 +1074,50 @@ public Channel createChannel(boolean transactional) {
 		public void close() {
 			if (CachingConnectionFactory.this.cacheMode == CacheMode.CONNECTION) {
 				synchronized (CachingConnectionFactory.this.connectionMonitor) {
-					if (!this.target.isOpen()
-							|| CachingConnectionFactory.this.idleConnections.size() >=
-									CachingConnectionFactory.this.connectionCacheSize) {
-						if (logger.isDebugEnabled()) {
-							logger.debug("Completely closing connection '" + this + "'");
-						}
-						if (this.target.isOpen()) {
-							RabbitUtils.closeConnection(this.target);
-						}
-						this.notifyCloseIfNecessary();
-						CachingConnectionFactory.this.openConnections.remove(this);
-						CachingConnectionFactory.this.openConnectionNonTransactionalChannels.remove(this);
-						CachingConnectionFactory.this.openConnectionTransactionalChannels.remove(this);
-					}
-					else {
-						if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+					/*
+					 * Only connectionCacheSize open idle connections are allowed.
+					 */
+					if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+						if (!this.target.isOpen()
+								|| countOpenIdleConnections() >= CachingConnectionFactory.this.connectionCacheSize) {
 							if (logger.isDebugEnabled()) {
-								logger.debug("Returning connection '" + this + "' to cache");
-							}
-							CachingConnectionFactory.this.idleConnections.add(this);
-							if (CachingConnectionFactory.this.connectionHighWaterMark
-									.get() < CachingConnectionFactory.this.idleConnections.size()) {
-								CachingConnectionFactory.this.connectionHighWaterMark
-										.set(CachingConnectionFactory.this.idleConnections.size());
+								logger.debug("Completely closing connection '" + this + "'");
 							}
+							destroy();
```

Comment:
Yes, the "cacheSize" now represents the maximum number of currently open idle channels. The actual number of entries can be larger than that (up to the new limit) but we'll never have more than the cache size open and idle.

This is to preserve the proxy object so the permits are always correct - I considered continuing to remove them but it adds more complexity to an already complex class.